upgrade mkldnn-bridge

gujinghui opened the pull request:
1. reduce the overhead of mkldnn-bridge itself
2. remove redundant code and useless APIs
3. provide new operators, including int8 inner_product,  ND permute/transpose, elem_add/mul, and etc.
4. improve inner_product to support io format weights without implicit reorder
5. add SoftMax support

gujinghui commented:
@pytorchbot retest this please

gujinghui commented:
@Jianhui-Li
@jgong5
@uyongw

gujinghui commented:
@pytorchbot retest this please

gujinghui commented:
all failures are caused by test environ.

> #!/bin/bash -eo pipefail
~/workspace/.circleci/scripts/setup_linux_system_environment.sh
/bin/bash: /home/circleci/workspace/.circleci/scripts/setup_linux_system_environment.sh: No such file or directory
Exited with code 1

gujinghui commented:
The failure is not related to this change.

> May 17 10:59:12 ======================================================================
May 17 10:59:12 FAIL: test_HalfTensor_tril_zero_stride (__main__.TestCuda)
May 17 10:59:12 ----------------------------------------------------------------------
May 17 10:59:12 Traceback (most recent call last):
May 17 10:59:12   File "/var/lib/jenkins/workspace/test/common_utils.py", line 338, in wrapper
May 17 10:59:12     method(*args, **kwargs)
May 17 10:59:12   File "test_cuda.py", line 648, in tmp
May 17 10:59:12     self.assertEqual(cpu_tensor, gpu_tensor, precision)
May 17 10:59:12   File "/var/lib/jenkins/workspace/test/common_utils.py", line 483, in assertEqual
May 17 10:59:12     assertTensorsEqual(x, y)
May 17 10:59:12   File "/var/lib/jenkins/workspace/test/common_utils.py", line 475, in assertTensorsEqual
May 17 10:59:12     self.assertLessEqual(max_err, prec, message)
May 17 10:59:12 AssertionError: tensor(nan, dtype=torch.float32) not less than or equal to 1e-05 : 
May 17 10:59:12

gujinghui commented:
@yinghai 
Could you help review this PR.

Thanks a lot.

yinghai commented:
We need to update internal ideep before accepting.

gujinghui commented:
@yinghai 
Could you help review this PR?
The MKLDNN is upgraded to v0.19.

Thanks a lot.

gujinghui commented:
rebased to latest code base

bddppq commented:
We have run into the this error when compiling mkl-dnn v0.19 from source:
```
cpu/ref_softmax.cpp:171:25: error: 'VML_ERRMODE_NOERR' was not declared in this scope
         vmsExp(n, a, r, VML_ERRMODE_NOERR);
                         ^~~~~~~~~~~~~~~~~
```

gujinghui commented:
@bddppq  looks like the mkl version is diff. could you help check which version is your mkl on your machine?

bddppq commented:
@gujinghui it's 2017.0.098

gujinghui commented:
@bddppq  @VitalyFedyunin

Should be fixed. Pls try again. Thanks.

bddppq commented:
@gujinghui Where did you put the fix? The release tag v0.19 in mkl-dnn repo is still pointing to the same commit.

yinghai commented:
`#define INTEL_MKL_VERSION 20170003`

gujinghui commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -122,9 +122,6 @@ IF(NOT TARGET mkldnn)
   MESSAGE("Failed to include MKL-DNN target")
   RETURN()
 ENDIF(NOT TARGET mkldnn)
-IF(MKL_FOUND)
```
@bddppq 
disabled this flag for MKLDNN, which caused the build failure.
This flag didn't impact on pytorch any more.

gujinghui commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -122,9 +122,6 @@ IF(NOT TARGET mkldnn)
   RETURN()
 ENDIF(NOT TARGET mkldnn)
-IF(MKL_FOUND)
-  TARGET_COMPILE_DEFINITIONS(mkldnn PRIVATE -DUSE_MKL)
```
@bddppq
disabled this flag for MKLDNN, which caused the build failure.
This flag didn't impact on pytorch any more.

bddppq replied:
Does disabling MKL have perf implications on MKLDNN?

gujinghui replied:
according to below table from MKLDNN, should be no perf change.
Anyway, we use jit impl.

  /* USE_MKL      USE_CBLAS       effect
   * -------      ---------       ------
   * yes          yes             use Intel(R) MKL CBLAS
   * yes          no              use jit
   * no           yes             system-dependent CBLAS
   * no           no              use jit
   */

gujinghui committed:
upgrade mkldnn-bridge

1. reduce the overhead of mkldnn-bridge itself
2. remove redundant code and useless APIs
3. provide new operators, including int8 inner_product,
   nD permute/transpose, ele_add/mul, and etc.
4. improve inner_product to support io format weights
   without implicit reorder
5. add softmax support

Signed-off-by: Gu, Jinghui <jinghui.gu@intel.com>

gujinghui commented:
@bddppq @yinghai 
Move softmax into mkldnn-bridge. Pls help review.

gujinghui commented:
@pytorchbot rebase this please

bddppq commented:
@gujinghui We have built mkldnn v0.19 from source (with `-DMKLDNN_USE_MKL=OFF`), and have seen significant perf drop when running resnext101 inference run. I think a better "fix" to the issue would be adding mkl version detection in mkldnn and fallback to old api call in environment with older version of mkl.

gujinghui commented:
@bddppq 
The MKLDNN v0.19 is reverted. And the USE_MKL flag is set as before.
Only ideep update in this PR now. Pls help review.

Per compatibility issue bwt v0.19 and old MKL, we're working on it now.
Will return to you if any update.

Thanks a lot.

bddppq commented:
@gujinghui Thanks!
Thought currently in this PR the mkl-dnn submodule (within the ideep submodule) is still updated.

gujinghui commented:
> @gujinghui Thanks!
> Thought currently in this PR the mkl-dnn submodule (within the ideep submodule) is still updated.

Yes. The trivial change is seen in ideep in this PR.
But the commit ID of mkl-dnn should be identical as before.
